homeassistant/components/fan/xiaomi_miio.py

```diff
@@ -74,45 +152,95 @@
+    SERVICE_SET_AUTO_DETECT_ON: {'method': 'async_set_auto_detect_on'},
+    SERVICE_SET_AUTO_DETECT_OFF: {'method': 'async_set_auto_detect_off'},
+    SERVICE_SET_LEARN_MODE_ON: {'method': 'async_set_learn_mode_on'},
+    SERVICE_SET_LEARN_MODE_OFF: {'method': 'async_set_learn_mode_off'},
+    SERVICE_RESET_FILTER: {'method': 'async_reset_filter'},
     SERVICE_SET_LED_BRIGHTNESS: {
         'method': 'async_set_led_brightness',
         'schema': SERVICE_SCHEMA_LED_BRIGHTNESS},
+    SERVICE_SET_FAVORITE_LEVEL: {
+        'method': 'async_set_favorite_level',
+        'schema': SERVICE_SCHEMA_FAVORITE_LEVEL},
+    SERVICE_SET_VOLUME: {
+        'method': 'async_set_volume',
+        'schema': SERVICE_SCHEMA_VOLUME},
+    SERVICE_SET_EXTRA_FEATURES: {
+        'method': 'async_set_extra_features',
+        'schema': SERVICE_SCHEMA_EXTRA_FEATURES},
+    SERVICE_SET_TARGET_HUMIDITY: {
+        'method': 'async_set_target_humidity',
+        'schema': SERVICE_SCHEMA_TARGET_HUMIDITY},
 }
 
 
 # pylint: disable=unused-argument
 @asyncio.coroutine
 def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
-    """Set up the air purifier from config."""
-    from miio import AirPurifier, DeviceException
-    if PLATFORM not in hass.data:
-        hass.data[PLATFORM] = {}
+    """Set up the miio fan device from config."""
+    from miio import Device, DeviceException
+    if DATA_KEY not in hass.data:
+        hass.data[DATA_KEY] = {}
 
     host = config.get(CONF_HOST)
     name = config.get(CONF_NAME)
     token = config.get(CONF_TOKEN)
+    model = config.get(CONF_MODEL)
 
     _LOGGER.info("Initializing with host %s (token %s...)", host, token[:5])
 
-    try:
+    if model is None:
+        try:
+            miio_device = Device(host, token)
+            device_info = miio_device.info()
+            model = device_info.model
+            _LOGGER.info("%s %s %s detected",
+                         model,
+                         device_info.firmware_version,
+                         device_info.hardware_version)
+        except DeviceException:
+            raise PlatformNotReady
+
+    if model.startswith('zhimi.airpurifier.'):
+        from miio import AirPurifier
         air_purifier = AirPurifier(host, token)
-
-        xiaomi_air_purifier = XiaomiAirPurifier(name, air_purifier)
-        hass.data[PLATFORM][host] = xiaomi_air_purifier
-    except DeviceException:
-        raise PlatformNotReady
-
-    async_add_devices([xiaomi_air_purifier], update_before_add=True)
+        device = XiaomiAirPurifier(name, air_purifier, model)
+    elif model.startswith('zhimi.humidifier.'):
+        from miio import AirHumidifier
+        air_humidifier = AirHumidifier(host, token)
+        device = XiaomiAirHumidifier(name, air_humidifier, model)
+    else:
+        _LOGGER.error(
+            'Unsupported device found! Please create an issue at '
+            'https://github.com/rytilahti/python.miio/issues '
```

Comment:
Typo in the URL, should this be reported to homeassistant though (considering that the backend library may already support the device)? Just a thought though, this is fine for me either way.

Reply:
It would be a feature request. That's why I don't want to mention homeassistant here. I added a link to my development repository now. So we can keep the python-miio repo clean for protocol issues.